src/map/itemdb.cpp

```diff
@@ -1197,49 +1190,51 @@ int itemdb_searchname_array(struct item_data** data, int size, const char *str)
 	return count;
 }
 
+std::shared_ptr<s_item_group_entry> get_random_itemsubgroup(std::shared_ptr<s_item_group_random> random) {
+	if (random == nullptr)
+		return nullptr;
+
+	for (size_t j = 0, max = random->data.size() * 3; j < max; j++) {
+		std::shared_ptr<s_item_group_entry> entry = util::umap_random(random->data);
+
+		if (entry->rate == 0 || rnd() % random->total_rate < entry->rate)	// always return entry for rate 0 ('must' item)
+			return entry;
+	}
+
+	return util::umap_random(random->data);
+}
+
 /**
 * Return a random group entry from Item Group
 * @param group_id
 * @param sub_group: 0 is 'must' item group, random groups start from 1 to MAX_ITEMGROUP_RANDGROUP+1
 * @return Item group entry or NULL on fail
 */
-struct s_item_group_entry *itemdb_get_randgroupitem(uint16 group_id, uint8 sub_group) {
-	struct s_item_group_db *group = (struct s_item_group_db *) uidb_get(itemdb_group, group_id);
-	struct s_item_group_entry *list = NULL;
-	uint16 qty = 0;
+std::shared_ptr<s_item_group_entry> ItemGroupDatabase::get_random_entry(uint16 group_id, uint8 sub_group) {
+	std::shared_ptr<s_item_group_db> group = this->find(group_id);
 
-	if (!group) {
-		ShowError("itemdb_get_randgroupitem: Invalid group id %d\n", group_id);
-		return NULL;
-	}
-	if (sub_group > MAX_ITEMGROUP_RANDGROUP+1) {
-		ShowError("itemdb_get_randgroupitem: Invalid sub_group %d\n", sub_group);
-		return NULL;
-	}
-	if (sub_group == 0) {
-		list = group->must;
-		qty = group->must_qty;
-	}
-	else {
-		list = group->random[sub_group-1].data;
-		qty = group->random[sub_group-1].data_qty;
+	if (group == nullptr) {
+		ShowError("get_random_entry: Invalid group id %hu\n", group_id);
+		return nullptr;
 	}
-	if (!qty) {
-		ShowError("itemdb_get_randgroupitem: No item entries for group id %d and sub group %d\n", group_id, sub_group);
-		return NULL;
+
+	if (group->random.empty() || group->random.count(sub_group) == 0) {
```

Comment:
Maybe split this into two error messages so that the user know what he messed up.